### PR TITLE
Align API with libuv

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request aims to align the libudx socket and stream API with the UDP and TCP APIs of libuv. Some of the current differences became especially notable when aligning the libudx bindings with the filesystem bindings, both for React Native.

### CODE REVIEW

**Comment by kasperisager** [comment]
I've limited this to the socket API as it seems like there's still some work to do on the streams API before aligning it with libuv makes sense.

**Comment by kasperisager** [comment]
@mafintosh Is there a reason for the choice of `const char *buf, size_t buf_len` instead of [`uv_buf_t`](http://docs.libuv.org/en/v1.x/misc.html?highlight=buf#c.uv_buf_t)?

**Comment by mafintosh** [comment]
no, only that the lifecycle has to be very explicit docs wise if we use uv_buf_t (ie do you need to keep that reference alive or is it copying the pointers internally)

**Comment by kasperisager** [comment]
libuv takes the simple approach and always passes `uv_buf_t` by value. It's a tiny struct after all.

**Comment by mafintosh** [comment]
Ya that works

**Comment by vtjnash** [comment]
> libuv takes the simple approach and always passes uv_buf_t by value

Drive-by comment from a libuv developer: the libuv library does not normally pass uv_buf_t by value. It nearly always takes in a pointer to an array of them which is ABI-compatible with the underlying syscall (for example, pwritev).

**Comment by kasperisager** [comment]
I missed that, thanks so much for the clarification! 🙏

**Comment by kasperisager** [comment]
I don't know enough about C to answer my next question: Is passing a single struct with a `sizeof()` `n` any different from passing two or more arguments with a combined `sizeof()` `n`? Won't the resulting stack frame have the exact same size?

If so, I'd guess that passing a `uv_buf_t` would be equivalent to passing a `char *` plus a `size_t`.

**Comment by vtjnash** [comment]
Some ABI (notably Win64) pass a `struct` by pointer, regardless of the size. Most have a cutoff of 1-2 qwords where they will use memory instead of registers. Some (notably x86) always use memory for argument passing. Generally the platform ABI is attempting to pick whichever is assumed to be fastest.

**Comment by kasperisager** [comment]
That's super useful information, thanks!

File: src/udx.c [hunk 1]
```diff
@@ -557,7 +557,7 @@ trigger_send_callback (udx_t *socket, udx_packet_t *pkt) {
     udx_stream_t *stream = pkt->ctx;
 
     stream->status |= UDX_STREAM_DESTROYED;
-    close_maybe(stream, UDX_ERROR_DESTROYED);
+    close_maybe(stream, 0);
```
Comment (kasperisager) [inline src/udx.c:560]: @mafintosh I wasn't entirely sure about this one. The JavaScript side didn't treat `UDX_ERROR_DESTROYED` as an error.

Comment (mafintosh) [reply src/udx.c:560]: Yea I think that's correct. It needs an error here, and unsure why I didn't before. There is still a happy path right for onclose somewhere? (ie something calls close_maybe(stream, 0) when everything is great)

Comment (kasperisager) [reply src/udx.c:560]: That happens here on `master`, yes:

https://github.com/mafintosh/libudx/blob/615860d0c0cf1f6102d07b36364ac1f91434d9b7/src/udx.c#L500
https://github.com/mafintosh/libudx/blob/615860d0c0cf1f6102d07b36364ac1f91434d9b7/src/udx.c#L515

Comment (mafintosh) [reply src/udx.c:560]: Let's try it! I'm pretty sure your fix is correct.

**Comment by kasperisager** [comment]
@mafintosh This is good to go from my end.

**Pull request merged**